Add basic Documentation for setting up on Google Cloud

Georgehe4 opened the pull request.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/mango-prb/515/
Test PASSed.

akmorrow13 commented on docs/cloud/google-cloud.rst:
```diff
@@ -0,0 +1,52 @@
+
+.. code:: bash
+
+    wget https://gist.githubusercontent.com/Georgehe4/6bb1c142a9f68f30f38d80cd9407120a/raw/9b903e3b8746ee8f25911fe98925b53e9777002f/mango_install.sh
```
Can we put this script in Mango scripts?

Georgehe4 replied:
Added

akmorrow13 commented on docs/cloud/google-cloud.rst:
```diff
@@ -0,0 +1,52 @@
+            gs://mango-initialization-bucket/mango_install.sh
+
+
+Once the above steps are completed, simply ssh into the master node to run Mango.
```
Do you have the scripts for running docker?

Georgehe4 replied:
Added

akmorrow13 commented on docs/index.rst:
```diff
@@ -31,6 +31,12 @@ variety of platforms.
    docker/docker-examples
 
+.. toctree::
+   :caption: Cloud
```
Google Cloud

Georgehe4 replied:
We might add an AWS Sections later?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/mango-prb/517/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/mango-prb/523/
Test PASSed.

akmorrow13 commented on scripts/google_cloud_docker_install.sh:
```diff
@@ -0,0 +1,15 @@
+# update the apt package index:
+sudo apt-get -y update
+# finally, install docker
+sudo apt-get -y install docker-ce
```
newline

Georgehe4 replied:
don

akmorrow13 commented on scripts/google_cloud_docker_run.sh:
```diff
@@ -0,0 +1,36 @@
+       -e HIVE_DIR=${HIVE_DIR} \
+       -e HIVE_CONF_DIR=${HIVE_CONF_DIR} \
+       -e SPARK_DIST_CLASSPATH="/usr/lib/hadoop/etc/hadoop:/usr/lib/hadoop/lib/*:/usr/lib/hadoop/.//*:/usr/lib/hadoop-hdfs/./:/usr/lib/hadoop-hdfs/lib/*:/usr/lib/hadoop-hdfs/.//*:/usr/lib/hadoop-yarn/lib/*:/usr/lib/hadoop-yarn/.//*:/usr/lib/hadoop-mapreduce/lib/*:/usr/lib/hadoop-mapreduce/.//*" \
+       --entrypoint=/opt/cgl-docker-lib/mango/bin/mango-notebook \
```
when you were accessing google storage through docker, did it need any access keys?

Georgehe4 replied:
Yes I still haven't resolved the access key issue among the spark workers yet for bam files in gs. For now, accessing hdfs works fine but gs still needs to be investigated.

akmorrow13 commented on scripts/google_cloud_mango_install.sh:
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env bash
```
let's put these in bin/GCE/<all the scripts>

Georgehe4 replied:
done

akmorrow13 commented on scripts/google_cloud_mango_install.sh:
```diff
@@ -0,0 +1,59 @@
+fi
+
+if [[ "${ROLE}" == 'Master' ]]; then
+    conda install jupyter
```
Why are you installing docker here? docker is installed in the container, right?

Georgehe4 replied:
Docker is not provided on a default gce vm.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/mango-prb/525/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/mango-prb/526/
Test PASSed.

akmorrow13 commented on docs/cloud/google-cloud.rst:
```diff
@@ -0,0 +1,71 @@
+
+    hdfs dfs -put /<local machime path> /<hdfs path>
+
+An example docker startup script is available in the Mango `scripts directory <https://github.com/bigdatagenomics/mango/blob/master/bin/gce/google_cloud_docker_run.sh>`__ for running mango notebook [run with root permissions to work with docker].
```
Can you give the code to copy this from docker so a user can run the script?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/mango-prb/533/
Test PASSed.

akmorrow13 commented on bin/gce/google_cloud_docker_run.sh:
```diff
@@ -0,0 +1,37 @@
+export SPARK_HOME=/usr/lib/spark
+export SPARK_CONF_DIR=/usr/lib/spark/conf
```
This file is specifically for mango notebook. we can pass a user variable here that states whether to run notebook or browser based on the --entrypoint flag

akmorrow13 replied:
separate into run-notebook.sh and run-browser.sh

akmorrow13 commented on bin/gce/google_cloud_mango_install.sh:
```diff
@@ -0,0 +1,60 @@
+    echo "Jupyter Notebook extensions installed!"
+fi
+
+pip install cigar
```
shouldn't cigar be installed in the virtual env?

Georgehe4 replied:
This is so cigar is present on the worker nodes (it is also installed in the venv).

What configurations have you been setting to port python packages over to the worker nodes on the cluster? It looks like the python binary for the worker nodes is not being overwritten by the flags sent to Spark, which leads to "cigar not found" issues if it is not installed this way.

akmorrow13 replied:
won't this be installed outside of the virtual env? Do you know why cigar is causing an issue and not other imports?

Georgehe4 replied:
Cigar will be installed in the virtual env, just not through this process.

Since the pyspark jobs depend on cigar in the worker nodes, the python distribution on the worker nodes need to have cigar present. I don't believe we use any other imports in pyspark for mango-notebook.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/mango-prb/545/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/mango-prb/547/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/mango-prb/555/
Test PASSed.

Georgehe4 commented:
Hey @akmorrow13, I've added an example file and fixed an issue with block data reads (https://gist.github.com/Georgehe4/043ae9079349ac7b05bee1e0b3b4b4fa) - I believe this should be good to go.

akmorrow13 commented on bin/gce/google_cloud_docker_run.sh:
```diff
@@ -0,0 +1,70 @@
+export HADOOP_HDFS=/usr/lib/hadoop-hdfs
+export HADOOP_YARN=/usr/lib/hadoop-yarn
+export HADOOP_MAPREDUCE=/usr/lib/hadoop-mapreduce
+export HIVE_DIR=/usr/lib/hive
```
Do you need hive export?

Georgehe4 replied:
Nope - removed

akmorrow13 commented on bin/gce/google_cloud_docker_run.sh:
```diff
@@ -0,0 +1,70 @@
+       -e HADOOP_CONF_DIR=${HADOOP_CONF_DIR} \
+       -e HIVE_DIR=${HIVE_DIR} \
+       -e HIVE_CONF_DIR=${HIVE_CONF_DIR} \
+       -e SPARK_DIST_CLASSPATH="/usr/lib/hadoop/etc/hadoop:/usr/lib/hadoop/lib/*:/usr/lib/hadoop/.//*:/usr/lib/hadoop-hdfs/./:/usr/lib/hadoop-hdfs/lib/*:/usr/lib/hadoop-hdfs/.//*:/usr/lib/hadoop-yarn/lib/*:/usr/lib/hadoop-yarn/.//*:/usr/lib/hadoop-mapreduce/lib/*:/usr/lib/hadoop-mapreduce/.//*" \
```
can you comment the use of this line?

Georgehe4 replied:
Added - this is necessary to specify the hadoop version (and other dependencies) to match the cluster environment in google cloud.

akmorrow13 commented on bin/gce/google_cloud_mango_install.sh:
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
```
can you simplify into 1 install script?

Georgehe4 replied:
What do you mean? This script will be automatically run upon cluster creation.

akmorrow13 commented on docs/cloud/google-cloud.rst:
```diff
@@ -0,0 +1,84 @@
+Running Mango on Google Cloud
+=============================
+
```
how do you make a google cloud cluster?

Georgehe4 replied:
refer to the section below, there's a step to 
`Create the Cloud Dataproc Cluster`

akmorrow13 commented on docs/cloud/google-cloud.rst:
```diff
@@ -0,0 +1,84 @@
+
+More information about available public datasets on Google cloud can be found `online <https://cloud.google.com/genomics/v1/public-data>`__
+
+More information on using the dataproc cluster's Spark interface is available through `Google Cloud documentation <https://cloud.google.com/dataproc/docs/concepts/accessing/cluster-web-interfaces>`__
```
is there an example notebook you can push and reference on a non-1000g datasource?

Georgehe4 replied:
I will make some changes to mango-google-cloud.ipynb to reference a non-1000g datasource

akmorrow13 commented on example-files/mango-google-cloud.ipynb:
```diff
@@ -0,0 +1,115 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "alignmentFile = \"gs://genomics-public-data/ftp-trace.ncbi.nih.gov/1000genomes/ftp/data/150140/alignment/150140.chrom20.ILLUMINA.bwa.CHM1.20131218.bam\""
```
what other datasets are hosted besides 1000g? I ask because 1000g is the only datasource on AWS, so it would be nice to reference a different one in the GCE examples

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/mango-prb/556/
Test PASSed.

akmorrow13 commented:
install.sh, run-notebook.sh, run-browser.sh

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/mango-prb/562/
Test PASSed.

akmorrow13 commented on bin/gce/run-submit.sh:
```diff
@@ -0,0 +1,68 @@
+set -ex
```
change file to run-browser

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/mango-prb/564/
Test PASSed.

akmorrow13 commented:
looks good @Georgehe4 can you rebase?

Georgehe4 commented:
Moved to #360.

Can you enable squash merges for this repo?
